src/GovUk.Education.ExploreEducationStatistics.Common/Services/BlobCacheService.cs

```diff
@@ -11,97 +11,100 @@
+        public async Task<TItem> GetItem<TItem>(
+            IBlobCacheKey cacheKey,
+            Func<Task<TItem>> itemSupplier)
+            where TItem : class
         {
             // Attempt to read blob from the cache container
-            var cachedEntity = await ReadFromCache<TEntity>(blobContainer, cacheKey);
+            var cachedEntity = await GetItem<TItem>(cacheKey);
             if (cachedEntity != null)
             {
                 return cachedEntity;
             }
 
             // Cache miss - invoke provider instead
-            var entity = await entityProvider();
+            var entity = await itemSupplier();
 
             // Write result to cache as a json blob before returning
-            await WriteToCache(blobContainer, cacheKey, entity);
+            await SetItem(cacheKey, entity);
             return entity;
         }
 
-        public async Task<Either<ActionResult, TEntity>> GetItem<TEntity>(
-            IBlobContainer blobContainer,
-            ICacheKey cacheKey,
-            Func<Task<Either<ActionResult, TEntity>>> entityProvider)
-            where TEntity : class
+        public async Task<Either<ActionResult, TItem>> GetItem<TItem>(
+            IBlobCacheKey cacheKey,
+            Func<Task<Either<ActionResult, TItem>>> itemSupplier)
+            where TItem : class
         {
             // Attempt to read blob from the cache container
-            var cachedEntity = await ReadFromCache<TEntity>(blobContainer, cacheKey);
+            var cachedEntity = await GetItem<TItem>(cacheKey);
+
             if (cachedEntity != null)
             {
                 return cachedEntity;
             }
 
             // Cache miss - invoke provider instead
-            return await entityProvider().OnSuccessDo(async entity =>
+            return await itemSupplier().OnSuccessDo(async entity =>
             {
                 // Write result to cache as a json blob before returning
-                await WriteToCache(blobContainer, cacheKey, entity);
+                await SetItem(cacheKey, entity);
             });
         }
 
-        private async Task<TEntity?> ReadFromCache<TEntity>(
-            IBlobContainer blobContainer,
-            ICacheKey cacheKey)
-            where TEntity : class
+        public async Task<TItem?> GetItem<TItem>(IBlobCacheKey cacheKey)
+            where TItem : class
+        {
+            return (TItem?) await GetItem(cacheKey, typeof(TItem));
+        }
+
+        public async Task<object?> GetItem(IBlobCacheKey cacheKey, Type targetType)
         {
+            var blobContainer = cacheKey.Container;
             var key = cacheKey.Key;
 
             // Attempt to read blob from the storage container
             try
             {
-                return await _blobStorageService.GetDeserializedJson<TEntity>(blobContainer, key);
+                return await _blobStorageService.GetDeserializedJson(cacheKey.Container, cacheKey.Key, targetType);
```

Comment:
I think one of the original limitations of this implementation which still exists here is we have no relation constraining the cache key and the connection which the BlobStorageService is using. For example the potential exists for the BlobCacheKey to use a private container name, but if the service is set up to use blob caching with a BlobStorageService wired with the PublicStorage connection string, the storage service will go ahead and create that container in the public account, rather than private as desired. So that could get messy, for example in the Admin if we try to use this for caching items in the private account, as well as needing to invalidate items in the public account.

Not sure if there's anything we can do here to prevent this.

I think overall getting rid of the divide would be best!

Reply:
Edit: think I can see now how you have implemented service name attribute to be used as a way of targeting the correct service. 🙂

Reply:
Yep indeed! You're right that it would be preferable to consolidate everything into a single account though to avoid any of the inherent public/private nastiness that we've seen over the last few weeks.